### ISSUE TITLE
Is NFS/Samba the only way to share files between host and guest?

### ISSUE THREAD

**Issue by jgonera** [mist64/xhyve#70]
I was wondering if I am missing something. Can I somehow mount the disk image both on guest and host? I suspect that this would cause concurrency issues even if it is possible.

**Comment by xez** [comment]
Network based solutions are probably the only options right now. I suggested `virtio-9p` as means of guest <-> host file sharing. But thats quite a bit of work.

**Comment by geoff-codes** [comment]
@jgonera osxfuse + [ext4fuse](https://github.com/gerard/ext4fuse) should allow you to mount R/W concurrently, though I can't vouch for it personally. Myself I use Paragon's commercial ExtFS for Mac,  which is quite fast and the write locks are good enough I've never seen a race condition or corruption.

**Comment by johanneswuerbach** [comment]
I think the main advantage of this driver vs virtualbox is that you don't need to install additional software. So while extfuse might work, it requires more software.

**Comment by geoff-codes** [comment]
@johanneswuerbach Good point... osxrefuse requires root and a codesigned kext. So I'll venture a different  "no" answer — there's also afp (netatalk/avahi), http/webdav, ftp, scp, netcat... :wink: :neckbeard:.

**Comment by galaxy001** [comment]
It should be possible to mimic the Parallel Desktop's solution that installs a driver to the guest and export filesystem to host.

If we export it to host under fuse, no other kext needed except for osxfuse itself.

------

@geoff-codes mentioned mount R/W for a disk image with write locks, I wonder whether the "write locks" means it can make both the virtual machine and host to write on same disk image ?

**Comment by jceel** [comment]
Hi guys,

I'm working on virtio-9p (so-called VirtFS) support for bhyve/xhyve. Right now I have almost complete 9p2000.u protocol support (except atime/mtime modification in Twstat).

You can check it out here: https://github.com/jceel/xhyve

Here's how to use it:
1. add `-s 3,virtio-9p,sharename=/host/path` to your xhyve arguments
2. boot Linux
3. `mount -t 9p -o trans=virtio -o version=9p2000.u sharename /mnt/guest/path`

Using 9p as root filesystem should work too.

This code is of course totally experimental and untested - I'm looking forward to your feeback!

**Comment by xez** [comment]
@jceel Awesome, great work!!!

**Comment by ailispaw** [comment]
Hi @jceel great work!!
I tried to use virtfs with your xhyve and it works, but I have one problem here.
When the owner of a file and a user in a VM have different uid, the user (except root) can't modify the file from the VM.
Is there any configuration for ACL mapping like `-mapall` option for NFS?

**Comment by jceel** [comment]
@ailispaw: You can try the following mount options: `access=any` (and optionally `dfltuid=<n>` if going to attach as non-root). That should make Linux do single attach as root/`dfltuid`.

Here's mount -t 9p options reference: https://www.kernel.org/doc/Documentation/filesystems/9p.txt

**Comment by ailispaw** [comment]
@jceel Thank you for the reply.
Yes. I have added `access=any` to it, but I can read files but write.
I will try `dfltuid=<n>`. Should `<n>` be a uid on host, shouldn't it?

**Comment by jceel** [comment]
@ailispaw: yes, it should refer to the host UID.

**Comment by ailispaw** [comment]
Kernel version: 4.4.1 with 
```
CONFIG_NET_9P=y
CONFIG_NET_9P_VIRTIO=y
CONFIG_9P_FS=y
CONFIG_9P_FS_POSIX_ACL=y
```

`501` is my uid in OSX 10.10.5 host.

```
$ sudo mkdir -p /mnt/host
$ sudo mount -t 9p -o trans=virtio -o version=9p2000.u -o access=any -o dfltuid=501 host /mnt/host
$ touch /mnt/host/aaa
touch: /mnt/host/aaa: Permission denied
```

Something missing?

**Comment by jceel** [comment]
Ah, I see what you're trying to do. There's no UID/GID translation mechanism (yet) and it turns out `dfltuid` works a bit differently than I thought:

- it's effective only when using legacy 9P2000 protocol (because original 9P2000 doesn't have concept of numeric uids/gids)
- it's a client uid

So, a very poor man's solution would be to use 9P2000 and map *everything* to uid specified by `dfltuid`. For example:

`mount -t 9p -o version=9p2000 -o trans=virtio -o access=any -o uname=root -o dftlduid=1000`

but with that approach, you'll lose ability to create symlinks/pipes/devices and ability see/modify file ownerships - basically everything in shared filesystem will be shown as owned by UID 1000. on the host side, modified files will be owned by root.

(to use `-o uname` option you need to update your sources, I've checked in support for it few minutes ago)

**Comment by ailispaw** [comment]
@jceel Thank you so much for the explanation.
I thought `security_mode=mapped` on the xhyve side was for that purpose.

**Comment by ailispaw** [comment]
@jceel I would like to let you know the result of `9p2000` case.
I tried `mount -t 9p -o version=9p2000 -o trans=virtio -o access=any -o uname=root -o dfltuid=1000 host /mnt/host`, but I got a xhyve panic as below.

```
$ sudo mkdir -p /mnt/host
$ sudo mount -t 9p -o version=9p2000 -o trans=virtio -o access=any -o uname=root -o dftlduid=1000 host /mnt/host
xhyve(5451,0x14468d000) malloc: *** error for object 0x7ffbd07006b0: pointer being freed was not allocated
     *** set a breakpoint in malloc_error_break to debug
                                                        ./xhyverun.sh: line 68:  5451 Abort trap: 6
```

**Comment by ailispaw** [comment]
@jceel Oh, I'm sorry I didn't update xhyve. I will test it again.

**Comment by ailispaw** [comment]
@jceel 
That solution works for me!!
```
$ sudo mount -t 9p -o version=9p2000 -o trans=virtio -o access=any -o uname=ailispaw -o dfltuid=1000 -o dfltgid=1000 host /mnt/host
```

Thanks. :)

**Comment by stv0g** [comment]
Do you plan to push those changes to @mist64's repo?

**Comment by galaxy001** [comment]
I tried jceel/xhyve with sysresccd which provides `fs/9p/9p.ko`. However, when I try to`mount -t 9p -o trans=virtio,version=9p2000.L hostshare /mnt/` after `modprobe 9p`, it says: 

    mount: special device hostshare does not exist

And `find /dev/ |grep hostshare` is empty.

My `lsmod` is:
```
% lsmod|grep -P '9p|virtio'
9p                     41056  0
fscache                47757  1 9p
9pnet_virtio           12537  0
9pnet                  58902  2 9p,9pnet_virtio
virtio_pci             12487  0
virtio_ring            14405  2 virtio_pci,9pnet_virtio
virtio                 13313  2 virtio_pci,9pnet_virtio
```

**Comment by jceel** [comment]
9p2000.L is not supported yet - try 9p2000.u.

**Comment by galaxy001** [comment]
I keep get this in `dmesg` each time run `mount -t 9p`:
```
9pnet_virtio: no channels available
```

I googled to https://bugs.launchpad.net/qemu/+bug/648128 and find 2 patch for `trans_virtio.c` and `mount.c`.

**Comment by jceel** [comment]
What did you pass to xhyve command line?

**Comment by galaxy001** [comment]
```
build/xhyve -A -m 1G -s 0:0,hostbridge -s 31,lpc -l com1,stdio -s 2:0,virtio-net -s 3,ahci-cd,/Users/Galaxy/Downloads/systemrescuecd-x86-4.7.2.iso -s 5,virtio-9p,hostshare=/Users/Galaxy/git/xhyve/t,ro -f kexec,systemrescuecd/rescue64,systemrescuecd/initram.igz,earlyprintk=serial console=ttyS0
```

**Comment by galaxy001** [comment]
@jceel Would you update a kernel binary in your repo for testing ?

If my CLI was correct, it should due to the systemrescuecd kernel cannot support even after `modprobe 9p`.
